src/security/builtin_plugins/authentication/src/auth_utils.c

```diff
@@ -0,0 +1,1530 @@
+        ddsrt_free(msg);
+        goto err_get_subject;
+    }
+
+    /* TODO: check if this is the correct format of the subject name: check spec */
+    X509_NAME_print_ex(bio, name, 0, XN_FLAG_RFC2253);
+
+    sz = (size_t)BIO_get_mem_data(bio, &pmem);
+    subject = ddsrt_malloc( sz + 1);
+
+    if (BIO_gets(bio, subject, (int32_t)sz + 1) < 0) {
+        char *msg = get_openssl_error_message();
+        DDS_Security_Exception_set(ex, DDS_AUTH_PLUGIN_CONTEXT, DDS_SECURITY_ERR_UNDEFINED_CODE, DDS_SECURITY_VALIDATION_FAILED, "X509_get_subject_name failed : %s", msg);
+        ddsrt_free(msg);
+        ddsrt_free(subject);
+        subject = NULL;
+    }
+
+    BIO_free(bio);
+
+    return subject;
+
+err_get_subject:
+    BIO_free(bio);
+err_bio_alloc:
+    return NULL;
+}
+
+dds_time_t
+get_certificate_expiry(
+    X509 *cert)
+{
+    dds_time_t expiry = DDS_TIME_INVALID;
+    ASN1_TIME *ans1;
+
+    assert(cert);
+
+    ans1 = X509_get_notAfter(cert);
+    if (ans1 != NULL) {
+        int days;
+        int seconds;
+        if (ASN1_TIME_diff(&days, &seconds, NULL, ans1) != 0) {
+            static const dds_duration_t secs_per_day = 86400;
+            dds_duration_t delta = ((dds_duration_t)seconds + ((dds_duration_t)days * secs_per_day)) * DDS_NSECS_IN_SEC;
+            expiry = dds_time() + delta;
+        }
+    }
+
+    return expiry;
+}
+
+int32_t
+get_subject_name_DER_encoded(
+        const X509 *cert,
+        unsigned char **buffer,
+        DDS_Security_SecurityException *ex)
+{
+    X509_NAME *name;
+    unsigned char *tmp = NULL;
+    int32_t sz = -1;
+
+    assert(cert);
+    assert(buffer);
+
+    name = X509_get_subject_name((X509 *)cert);
+    if (name) {
+        sz = i2d_X509_NAME(name, &tmp);
+        if (sz > 0) {
+            *buffer = ddsrt_malloc((size_t)sz);
+            memcpy(*buffer, tmp, (size_t)sz);
+            OPENSSL_free(tmp);
+        } else if (sz < 0) {
+            char *msg = get_openssl_error_message();
+            DDS_Security_Exception_set(ex, DDS_AUTH_PLUGIN_CONTEXT, DDS_SECURITY_ERR_UNDEFINED_CODE, DDS_SECURITY_VALIDATION_FAILED, "i2d_X509_NAME failed : %s", msg);
+            ddsrt_free(msg);
+        }
+    } else {
+        char *msg = get_openssl_error_message();
+        DDS_Security_Exception_set(ex, DDS_AUTH_PLUGIN_CONTEXT, DDS_SECURITY_ERR_UNDEFINED_CODE, DDS_SECURITY_VALIDATION_FAILED, "X509_get_subject_name failed : %s", msg);
+        ddsrt_free(msg);
+    }
+
+    return sz;
+}
+
+
+static DDS_Security_ValidationResult_t
+check_key_type_and_size(
+    EVP_PKEY *key,
+    int isPrivate,
+    DDS_Security_SecurityException *ex)
+{
+    DDS_Security_ValidationResult_t result = DDS_SECURITY_VALIDATION_OK;
+    const char *sub = isPrivate ? "private key" : "certificate";
+
+    assert(key);
+
+    switch (EVP_PKEY_id(key)) {
+    case EVP_PKEY_RSA:
+        if (EVP_PKEY_bits(key) != 2048) {
```

Comment:
Why the key size restrictions?

Reply:
It comes from the spec: 

> The public key of the CA shall be either a 2048-bit RSA key [44] or else a 256-bit Elliptic Curve Key for the prime256v1 curve [41], also known as the NIST P-256 curve [42].

Reply:
Your argument is valid, but IMHO the spec is wrong and should have read something like (for the RSA key, analogous for the EC one): "public key shall be at least 2048-bit RSA; implementations are not required to support longer than 2048-bit RSA; implementations should support at least 2048-bit and 4096-bit" But whatever ...

Reply:
By the way, increasing the key size dramatically slows down the signing and validation process.